File: 101-is-symmetric/solution.go
//這版本是拿java版本來改的, 還是鏡像遍歷, 但是改用for迴圈來寫
package main

import "fmt"
import "gopractice/datastructure/Tree"
type TreeNode = Tree.TreeNode
func isSymmetric(root *TreeNode) bool {
	if root == nil{ //理論上不會走到這裡
		return  false
	}
	leftside := leftSide(root.Left)
	rightside := rightSide(root.Right)
	fmt.Println("leftside = ", leftside)
	fmt.Println("rightside = ", rightside)
	
	if len(leftside) != len(rightside){ //要先檢查長度是否相等, 否則會報錯
		return false
	}
	for i := 0; i < len(leftside); i++ {
		if leftside[i] != rightside[i] {
			return false
		}
	}
	return true
}

func leftSide(root *TreeNode) []interface{} {
	if root == nil{
		return []interface{}{}
	}
	var result []interface{}
	queue := []*TreeNode{root}
	for len(queue) > 0 {
		curNode := queue[0]
		queue = queue[1:] // 彈出
		if curNode != nil {
			result = append(result, curNode.Val)
			queue = append(queue, curNode.Left)
			queue = append(queue, curNode.Right)
		} else {
			result = append(result, nil)
		}
	}
	return result
}

func rightSide(root *TreeNode) []interface{} {
	if root == nil{
		return []interface{}{}
	}
	var result []interface{}
	queue := []*TreeNode{root}
	for len(queue) > 0 {
		curNode := queue[0]
		queue = queue[1:] // 彈出
		if curNode != nil {
			result = append(result, curNode.Val)
			queue = append(queue, curNode.Right)
			queue = append(queue, curNode.Left)
		} else {
			result = append(result, nil)
		}
	}
	return result
}
func main(){
	nums := []any {1,0}
	root := Tree.BuildLevelOrderTree(nums)
	fmt.Println("Ans = ", isSymmetric(root))
}
/*
建議先看過102題, 了解層序遍歷是如何運作的
102-level-order

執行代碼
go run ./101-is-symmetric
*/

File: 101-is-symmetric/solution.py
#K神題解
import sys
from pathlib import Path
sys.path.insert(0, str(Path(__file__).parent.parent))
from typing import Optional
from datastructure  import TreeNode
from datastructure import TreeDebugger

class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        if root is None: #基本不會走到這裡
            return False
        return self.recursion(root.left, root.right)
    def recursion(self, L: Optional[TreeNode], R: Optional[TreeNode]) -> bool:
        if not L and not R:
            return True
        elif (not L or not R) or (L.val != R.val):
            return False
        return self.recursion(L.left, R.right) and self.recursion(L.right, R.left)
if __name__ == "__main__":
    nums = [1,2,2,3,4,4,3]
    root = TreeDebugger().build_levelorder_tree(nums)
    print("Ans = ", Solution().isSymmetric(root))

    """
    官解
    https://leetcode.cn/problems/symmetric-tree/solutions/268109/dui-cheng-er-cha-shu-by-leetcode-solution/
    K神題解
    https://leetcode.cn/problems/symmetric-tree/solutions/2361627/101-dui-cheng-er-cha-shu-fen-zhi-qing-xi-8oba/

    兩個版本都看完, 再用自己的理解寫一遍
    
    
    recursion(2, 2)  ← 初始调用
        ├─ recursion(3, 3)  ← L.left vs R.right
        │   ├─ recursion(None, None) → True
        │   └─ recursion(None, None) → True
        │   └─ 返回 True
        │
        └─ recursion(4, 4)  ← L.right vs R.left
            ├─ recursion(None, None) → True
            └─ recursion(None, None) → True
            └─ 返回 True
        └─ 返回 True (True AND True)

    相關題目
    100-is-same-tree
    """